lib/dao.js

```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
 };
 
 /**
+ * Replace set of attributes.
+ * Performs validation before replacing.
+ *
+ * @trigger `validation`, `save` and `update` hooks
+ * @param {Object} data Data to replace
+ * @param {Object} [options] Options for replace
+ * @param {Function} cb Callback function called with (err, instance)
+ */
+
+DataAccessObject.prototype.replace = function replace(data, options, cb) {
+  var self = this;
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+
+  if (options === undefined && cb === undefined) {
+    if (typeof data === 'function') {
+      cb = data;
+      data = undefined;
+    }
+  } else if (cb === undefined) {
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+          'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var inst = this;
+  var Model = this.constructor;
+  var connector = inst.getConnector();
+  assert(typeof connector.replace === 'function',
+          'replace() must be implemented by the connector');
+
+  if (isPKMissing(Model, cb))
+    return cb.promise;
+
+
+  var strict = this.__strict;
+  var model = Model.modelName;
+  var hookState = {};
+
+  // Convert the data to be plain object so that update won't be confused
+  if (data instanceof Model) {
+    data = data.toObject(false);
+  }
+  data = removeUndefined(data);
+
+  // Make sure id(s) cannot be changed
+  var idNames = Model.definition.idNames();
+  for (var i = 0, n = idNames.length; i < n; i++) {
+    var idName = idNames[i];
+    if (data[idName] !== undefined && !idEquals(data[idName], inst[idName])) {
+      var err = new Error('id property (' + idName + ') ' +
+              'cannot be updated from ' + inst[idName] + ' to ' + data[idName]);
+      err.statusCode = 400;
+      process.nextTick(function() {
+        cb(err);
+      });
+      return cb.promise;
+    }
+  }
+  var context = {
+    Model: Model,
+    instance: inst,
+    currentInstance: inst,
+    hookState: hookState,
+    options: options
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    if (strict) {
+      self.strictChk(strict, data, inst, strictChkCB);
+    } else {
+      strictChkCB(null, data);
+    }
+
+    function strictChkCB(err, data) {
+      if (err) return cb(err);
+      // update instance's properties    
+      inst.setAttributes(data);
+
+      inst.isValid(function(valid) {
```

Comment:
Please take into account `Model.settings.automaticValidation` (see the usage elsewhere).